### PR TITLE
avoid fread on directories and unencrypted files

### PR DESCRIPTION
**Pull request by jknockaert**
Reworking the logic in order to first check the filecache and only then reading the fileheader.
This in order to solve #21578.
Signed-off-by: Jasper Knockaert jasper@knockaert.nl

### CODE REVIEW

**Comment by jknockaert** [comment]
@nextcloud/encryption Pls review. Also check the discussion of the issue I posted in #21578. That said, I didn't yet find out why this function is evaluated very often for keyfiles.

**Comment by jknockaert** [comment]
Signed-off-by: Jasper Knockaert jasper@knockaert.nl

**Comment by jknockaert** [comment]
Can anyone help with the testing? It is my understanding that in the tests the call to the cache was never evaluated, so I guess that should somewhere be caught now that it comes first.

**Comment by jknockaert** [comment]
OK, with all checks passed now I think this one is ready for review. As for the `getHeader` function I'm pretty sure I fully understand the code and I think my implementation is fine; but still then I don't have a full overview of all use cases of this function so I may miss out on a corner case. As for the tests I somehow got it working; not sure though if I fully understand what I did here. I couldn't do any meaningful benchmarking so I can't say much on the impact of the changes introduced. I wouldn't expect it to matter much.

**Comment by axheli** [comment]
@jknockaert I applied the below fix  and im not getting the errors anymore  

image

**Comment by RedKage** [comment]
I dont know what this is doing, nor am I a nextcloud maintainer.
However I think that IF can be simplified and be made more readable with

```
    if (!isset($result[Util::HEADER_ENCRYPTION_MODULE_KEY] && (!empty($result) || $exists)) {
        $result[Util::HEADER_ENCRYPTION_MODULE_KEY] = 'OC_DEFAULT_MODULE';
    }
```

**Comment by jknockaert** [comment]
@RedKage Thanks; fixed.

**Comment by jknockaert** [comment]
@ChristophWurst Any suggestion who to invite for a review?

**Comment by jknockaert** [comment]
@nextcloud/encryption Can anyone review? This is a fairly minor patch that fixes errors with php7.4. It would be good to have it included in the next point release of nextcloud.

**Comment by moverduinsecov** [comment]
Have applied changes manually, and so far everything looks fine.

**Comment by jknockaert** [comment]
Can anyone review? This is a minor (and straightforward) patch that fixes errors with php7.4. It would be good to have it included in the next point release of nextcloud.

**Comment by axheli** [comment]
@ChristophWurst Can you please have someone review this?

**Comment by axheli** [comment]
@rullzer anyway we can get this approved? Please :)

**Comment by axheli** [comment]
So who's going to approve this?

**Review by icewind1991 — APPROVED** [review]
I believe that this should be fine, but I haven't looked into the encryption logic for a long time

**Comment by jknockaert** [comment]
Can anyone review pls? This PR was submitted three months ago and is limited to a reordering of logic in a single function.

**Comment by axheli** [comment]
This is crazy we’re just asking for review and approval why dose it take so so long for a simple review.

**Comment by talesam** [comment]
I saw that it has not yet been added to the nextcloud, how do I add it manually?

File: lib/private/Files/Storage/Wrapper/Encryption.php [hunk 1]
```diff
@@ -929,21 +929,20 @@ protected function getHeader($path) {
 			$path = $realFile;
```
Comment (vitormattos) [inline lib/private/Files/Storage/Wrapper/Encryption.php:929]: Does it make sense to do something if the file doesn't exist?
it seems to me that here it should be `if (!$exists)` and have a `return [];`
This simplifies the code below

**Review by vitormattos — CHANGES_REQUESTED** [review]

**Comment by vitormattos** [comment]
@icewind1991 

> I believe that this should be fine, but I haven't looked into the encryption logic for a long time

If the test passes, what's wrong? :-)

File: lib/private/Files/Storage/Wrapper/Encryption.php (hunk 1 above)
Comment (jknockaert) [reply lib/private/Files/Storage/Wrapper/Encryption.php:929]: I didn't change anything in this respect. This is pretty low level code, so I prefer not to change it too much. It's hard to predict the context of all the different cases in which this code is invoked.

Comment (vitormattos) [reply lib/private/Files/Storage/Wrapper/Encryption.php:929]: Without tests covering the scenarios it is difficult to review.

Comment (jknockaert) [reply lib/private/Files/Storage/Wrapper/Encryption.php:929]: Unfortunately the tests are not covering every use scenario, far from it. Otherwise the php error on deleting a folder would have been caught by a test. Unfortunately I only know the lower-level part of the encryption code in Nextcloud; I have little experience with the filesystem code.

Comment (vitormattos) [reply lib/private/Files/Storage/Wrapper/Encryption.php:929]: The test you changed just made the test not break, there are no new data to cover possible scenarios for the change.

The test method testGetHeaderAddLegacyModule has a @dataProvider with a list of possible scenarios and outputs (dataTestGetHeaderAddLegacyModule).

It is good practice to include the scenario that covers the identified bug and the scenarios that cover the changes you made to the code.

On the comment I made about the return [] when the file does not exist, looking at the data from the test scenarios, the only scenario that has $exists = false returns an empty array. It does not make sense that the file does not exist and the method returns a header for a file that does not exist.

Comment (jknockaert) [reply lib/private/Files/Storage/Wrapper/Encryption.php:929]: As for the testing you're correct, but I don't know how to write such a new test. It should take into account that only php 7.4 fails. And I am not sure what the filecache for a folder should be (I have little understanding of the filecache specification).
As for the `return []` comment I prefer to stick to the current implementation. There seems to be a case where `$realFile` does not exist but `$path` does; see e.g. https://github.com/nextcloud/server/blob/40fde94b4d019f5c1914225d5be6854241abeb9c/lib/private/Files/Storage/Wrapper/Encryption.php#L398
Also in `getHeaderSize()` it is implemented such that `$path` is updated if `$realFile` exists. I am not sure which scenario is covered here. It may be something to do with versioning, or with files on an external (mounted) storage. Perhaps we should leave it for a separate PR if there is a need to fix this.

Comment (vitormattos) [reply lib/private/Files/Storage/Wrapper/Encryption.php:929]: I saw the tests better and also didn't identify new data scenarios.

**Review by vitormattos — APPROVED** [review]
It seems to be ok, there seems to be no need to create new test scenarios.

**Comment by jknockaert** [comment]
Thanks @vitormattos and @icewind1991 for the reviews.

**Comment by axheli** [comment]
@jknockaert 
@vitormattos 

Just updated to 20.0.9 and got this error dose this have anything to do with this ?

Error: Declaration of OC\Files\Stream\Encryption::wrapSource($source, $context, $protocol, $class, $mode = 'r+') should be compatible with Icewind\Streams\WrapperHandler::wrapSource($source, $context = Array, $protocol = NULL, $class = NULL) at /var/www/nextcloud/lib/private/Files/Stream/Encryption.php#201



Error: Declaration of OC\Files\Stream\Encryption::loadContext($name) should be compatible with Icewind\Streams\Wrapper::loadContext($name = NULL) at /var/www/nextcloud/lib/private/Files/Stream/Encryption.php#224
--

**Comment by vitormattos** [comment]
It is in the same class but does not seem to be related to this pull request.

Looking at the log quickly, without a thorough analysis, the method changed in this pull request is not in the callstack of the log you posted.

**Comment by axheli** [comment]
@vitormattos thank you for the quick reply. Im not sure what happen as soon as i update to 20.0.9 even if i open up a file or scroll throw the phones tab log files goes crazy. what do you suggest next for me to do ? 

image

**Comment by vitormattos** [comment]
I suggest sending a new pull request solving this problem :)

The solution is already described in the log, it is necessary to implement, check the impacts, review the unit tests to ensure they are ok and send the pull request.

**Comment by axheli** [comment]
I just did! i tagged you in it so maybe we can have someone look at it quicker :)

**Comment by vitormattos** [comment]
You have opened an issue. Submit the pull request with the solution. The correction is simple, but it is necessary to check the impacts well.

**Comment by axheli** [comment]
@vitormattos i would not know the solutions i don't code. maybe @jknockaert knows it but i dont

**Comment by jknockaert** [comment]
@axheli Pls tag me. I may have time to have a look at it over the weekend.
That said, I think the filesystem code is in some need for a more thorough review and possibly rewrite.

**Comment by vitormattos** [comment]
> I just did! i tagged you in it so maybe we can have someone look at it quicker :)

@jknockaert #26464

**Comment by axheli** [comment]
@jknockaert Thank you so much :)

**Comment by andyxheli** [comment]
Still an issue on Nextcloud 21.0.2 @LukasReschke can we please add this fix ?

**Comment by andyxheli** [comment]
Hello everyone is anybody willing to  add this fix to the Next release 21.0.4 ?
